# Regression Exercises

<hr style="border-top: 10px groove cyan; margin-top: 1px; margin-bottom: 1px"></hr>

## Exercises I - Required
Let's review the steps we take at the beginning of each new module.

1. Create a new repository named regression-exercises in your GitHub; all of your Regression work will be housed here.
2. Clone this repository within your local codeup-data-science directory.
3. Create a .gitignore and make sure your list of 'files to ignore' includes your env.py file.
4. Ceate a README.md file that outlines the contents and purpose of your repository.
5. Add, commit, and push these two files.
6. Now you can add your env.py file to this repository to access the Codeup database server.
7. For these exercises, you will create wrangle.ipynb and wrangle.py files to hold necessary functions.
8. As always, add, commit, and push your work often.

In [1]:
import pandas as pd
import numpy as np
import acquire
import wrangle as w

<hr style="border-top: 10px groove cyan; margin-top: 1px; margin-bottom: 1px"></hr>

## Exercises II - Required
Let's set up an example scenario as perspective for our regression exercises using the Telco dataset.

As a customer analyst for Telco, you want to know who has spent the most money with the company over their lifetime. You have monthly charges and tenure, so you think you will be able to use those two attributes as features to estimate total charges. You need to do this within an average of $5.00 per customer.

In these exercises, you will complete the first step toward the above goal: acquire and prepare the necessary Telco data from the telco_churn database in the Codeup database server.

1. Acquire customer_id, monthly_charges, tenure, and total_charges from the telco_churn database for all customers with a 2-year contract.
2. Using your acquired Telco data, walk through the summarization and cleaning steps in your wrangle.ipynb file like we did above. You may handle the missing values however you feel is appropriate and meaningful; remember to document your process and decisions using markdown and code commenting where helpful.
3. End with a wrangle.py file that contains the necessary functions to automate your process from acquiring the data to returning a cleaned dataframe with no missing values. Name your final function wrangle_telco.

In [25]:
query = "SELECT customer_id, monthly_charges, tenure, total_charges FROM customers WHERE `contract_type_id` = 3"

In [26]:
df = acquire.get_telco_data(query)

In [27]:
df.head()

customer_id  monthly_charges  tenure total_charges
0  0013-SMEOE           109.70      71       7904.25
1  0014-BMAQU            84.65      63        5377.8
2  0016-QLJIS            90.45      65        5957.9
3  0017-DINOC            45.20      54       2460.55
4  0017-IUDMW           116.80      72       8456.75

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1695 entries, 0 to 1694
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   customer_id      1695 non-null   object 
 1   monthly_charges  1695 non-null   float64
 2   tenure           1695 non-null   int64  
 3   total_charges    1695 non-null   object 
dtypes: float64(1), int64(1), object(2)
memory usage: 53.1+ KB


In [29]:
df.describe().T

count       mean        std   min     25%    50%    75%  \
monthly_charges  1695.0  60.770413  34.678865  18.4  24.025  64.35  90.45   
tenure           1695.0  56.735103  18.209363   0.0  48.000  64.00  71.00   

                    max  
monthly_charges  118.75  
tenure            72.00

#### Takeaways
- total_charges is an object
- don't appear to be any blanks off the bat
- I know there are blanks in total_charges

In [30]:
# do value counts to see 10 spaces
df.total_charges.value_counts()

           10
1161.75     2
5714.2      2
3533.6      2
5682.25     2
           ..
2016.3      1
1177.95     1
3616.25     1
854.8       1
225.55      1
Name: total_charges, Length: 1678, dtype: int64

In [31]:
# see what the blank total charges look like
df[df.total_charges == ' ']

customer_id  monthly_charges  tenure total_charges
234   1371-DWPAZ            56.05       0              
416   2520-SGTTA            20.00       0              
453   2775-SEFEE            61.90       0              
505   3115-CZMZD            20.25       0              
524   3213-VVOLG            25.35       0              
678   4075-WKNIU            73.35       0              
716   4367-NUYAO            25.75       0              
726   4472-LVYGI            52.55       0              
941   5709-LVOEQ            80.85       0              
1293  7644-OMVMY            19.85       0

#### Dealing with the blanks (the whitespaces)
- they seem to have all tenure of 0 
- can replace total_charges with 0 because these people seem to have not been here for a month yet
- they haven't paid their first bill
- will replace total_charges with a 0 then convert to float

In [32]:
df['total_charges'] = df['total_charges'].str.replace(' ', '0').astype('float')

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1695 entries, 0 to 1694
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   customer_id      1695 non-null   object 
 1   monthly_charges  1695 non-null   float64
 2   tenure           1695 non-null   int64  
 3   total_charges    1695 non-null   float64
dtypes: float64(2), int64(1), object(1)
memory usage: 53.1+ KB


In [34]:
df.describe().T

count         mean          std   min       25%      50%  \
monthly_charges  1695.0    60.770413    34.678865  18.4    24.025    64.35   
tenure           1695.0    56.735103    18.209363   0.0    48.000    64.00   
total_charges    1695.0  3706.934336  2579.517834   0.0  1269.675  3593.80   

                     75%      max  
monthly_charges    90.45   118.75  
tenure             71.00    72.00  
total_charges    5988.80  8672.45

In [38]:
def wrangle_telco():
    # get dataframe using get_telco_data function
    df = get_telco_data()
    # deal with space values (turn to 0s)
    df['total_charges'] = df['total_charges'].str.replace(' ', '0').astype('float')
    return df
# this is now in the wrangle.py file along with the aquire functions nessiscary 

In [2]:
# import wrangle to test
import wrangle

In [4]:
# test function
test_df = wrangle.wrangle_telco()

In [5]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   customer_id      7043 non-null   object 
 1   monthly_charges  7043 non-null   float64
 2   tenure           7043 non-null   int64  
 3   total_charges    7043 non-null   float64
dtypes: float64(2), int64(1), object(1)
memory usage: 220.2+ KB


In [7]:
test_df.shape

(7043, 4)

In [10]:
test_df.describe().T

count         mean          std    min     25%      50%  \
monthly_charges  7043.0    64.761692    30.090047  18.25   35.50    70.35   
tenure           7043.0    32.371149    24.559481   0.00    9.00    29.00   
total_charges    7043.0  2279.734304  2266.794470   0.00  398.55  1394.55   

                     75%      max  
monthly_charges    89.85   118.75  
tenure             55.00    72.00  
total_charges    3786.60  8684.80

In [12]:
test_df.total_charges.value_counts()

0.00       11
20.20      11
19.75       9
20.05       8
19.90       8
           ..
1756.60     1
4138.05     1
1401.15     1
6029.90     1
675.60      1
Name: total_charges, Length: 6531, dtype: int64

<hr style="border-top: 10px groove blue; margin-top: 1px; margin-bottom: 1px"></hr>

## Exercises III - Challenge
Let's set up an example scenario as perspective for our regression exercises using the Zillow dataset.

As a Codeup data science graduate, you want to show off your skills to the Zillow data science team in hopes of getting an interview for a position you saw pop up on LinkedIn. You thought it might look impressive to build an end-to-end project in which you use some of their Kaggle data to predict property values using some of their available features; who knows, you might even do some feature engineering to blow them away. **Your goal is to predict the values of single unit properties using the obervations from 2017**.

In these exercises, you will complete the first step toward the above goal: acquire and prepare the necessary Zillow data from the zillow database in the Codeup database server.

1. Acquire bedroomcnt, bathroomcnt, calculatedfinishedsquarefeet, taxvaluedollarcnt, yearbuilt, taxamount, and fips from the zillow database for all 'Single Family Residential' properties.
2. Using your acquired Zillow data, walk through the summarization and cleaning steps in your wrangle.ipynb file like we did above. You may handle the missing values however you feel is appropriate and meaninful; remember to document your process and decisions using markdown and code commenting where helpful.
3. Store all of the necessary functions to automate your process from acquiring the data to returning a cleaned dataframe witn no missing values in your wrangle.py file. Name your final function wrangle_zillow.

In [14]:
def get_zillow_data():
    '''
    This function reads in Zillow data from Codeup database, writes data to
    a csv file if a local file does not exist, and returns a df.
    '''
    sql_query = ''' SELECT bedroomcnt, bathroomcnt, calculatedfinishedsquarefeet, taxvaluedollarcnt, yearbuilt, taxamount, fips
    FROM properties_2017
    WHERE propertylandusetypeid = 261;
    '''

    if os.path.isfile('zillow_data.csv'):
        
        # If csv file exists read in data from csv file.
        df = pd.read_csv('zillow_data.csv', index_col=0)
        
    else:
        
        # Read fresh data from db into a DataFrame
        df = pd.read_sql(sql_query, get_db_url('zillow'))
        
        # Cache data
        df.to_csv('zillow_data.csv')
    return df

In [2]:
df = w.get_zillow_data()

/opt/homebrew/anaconda3/lib/python3.8/site-packages/numpy/lib/arraysetops.py:583: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


In [4]:
df.head()

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  taxvaluedollarcnt  \
0         0.0          0.0                           NaN            27516.0   
1         0.0          0.0                           NaN               10.0   
2         0.0          0.0                           NaN               10.0   
3         0.0          0.0                           NaN             2108.0   
4         4.0          2.0                        3633.0           296425.0   

   yearbuilt  taxamount    fips  
0        NaN        NaN  6037.0  
1        NaN        NaN  6037.0  
2        NaN        NaN  6037.0  
3        NaN     174.21  6037.0  
4     2005.0    6941.39  6037.0

In [5]:
# interesting note, this doesn't show the non-null type count like other times I've run .info()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2152863 entries, 0 to 2152862
Data columns (total 7 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   bedroomcnt                    float64
 1   bathroomcnt                   float64
 2   calculatedfinishedsquarefeet  float64
 3   taxvaluedollarcnt             float64
 4   yearbuilt                     float64
 5   taxamount                     float64
 6   fips                          float64
dtypes: float64(7)
memory usage: 131.4 MB


In [11]:
df.isnull().any()

bedroomcnt                       True
bathroomcnt                      True
calculatedfinishedsquarefeet     True
taxvaluedollarcnt                True
yearbuilt                        True
taxamount                        True
fips                            False
dtype: bool

In [7]:
# I saw this on the afore mentioned kaggle site. This is the credit that author gave.
# credit: https://www.kaggle.com/willkoehrsen/start-here-a-gentle-introduction. 
# One of the best notebooks on getting started with a ML problem.

def missing_values_table(df):
    '''
    this function takes a dataframe as input and will output metrics for missing values, and the percent of that column that has missing values
    '''
        # Total missing values
        mis_val = df.isnull().sum()
        
        # Percentage of missing values
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        
        # Make a table with the results
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        
        # Rename the columns
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        
        # Sort the table by percentage of missing descending
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        
        # Print some summary information
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"      
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        
        # Return the dataframe with missing information
        return mis_val_table_ren_columns

In [8]:
missing_values_table(df)

Your selected dataframe has 7 columns.
There are 6 columns that have missing values.


Missing Values  % of Total Values
yearbuilt                               9337                0.4
calculatedfinishedsquarefeet            8484                0.4
taxamount                               4442                0.2
taxvaluedollarcnt                        493                0.0
bedroomcnt                                11                0.0
bathroomcnt                               11                0.0

In [13]:
# see where the nulls are
df.bedroomcnt.value_counts(dropna=False)

3.0     964298
4.0     634289
2.0     335473
5.0     150866
6.0      25166
1.0      23166
0.0      13187
7.0       4807
8.0       1107
9.0        291
10.0       121
11.0        34
13.0        16
12.0        12
NaN         11
14.0         7
15.0         6
18.0         3
16.0         2
25.0         1
Name: bedroomcnt, dtype: int64

In [19]:
# Create list of columns
cols = list(df.columns)

In [20]:
cols

['bedroomcnt',
 'bathroomcnt',
 'calculatedfinishedsquarefeet',
 'taxvaluedollarcnt',
 'yearbuilt',
 'taxamount',
 'fips']

In [23]:
for col in cols:
    print(f'------{col}-------')
    print(df[col].value_counts(dropna=False))

------bedroomcnt-------
3.0     964298
4.0     634289
2.0     335473
5.0     150866
6.0      25166
1.0      23166
0.0      13187
7.0       4807
8.0       1107
9.0        291
10.0       121
11.0        34
13.0        16
12.0        12
NaN         11
14.0         7
15.0         6
18.0         3
16.0         2
25.0         1
Name: bedroomcnt, dtype: int64
------bathroomcnt-------
2.00     943589
3.00     422841
1.00     414324
2.50     142981
4.00      82155
1.50      31211
3.50      28518
5.00      28362
4.50      19506
0.00      13027
6.00      10747
5.50       6217
7.00       4394
8.00       1692
6.50       1333
9.00        713
7.50        384
10.00       325
11.00       146
8.50        110
12.00        73
9.50         50
13.00        39
14.00        25
15.00        17
0.50         16
10.50        14
16.00        12
NaN          11
18.00         8
20.00         6
17.00         4
1.75          3
12.50         3
11.50         3
14.50         1
32.00         1
19.50         1
19.00       

#### Takeaways
['bedroomcnt',
 'bathroomcnt',
 'fips'] 
 
 - These columns are helpful for value counts
 
 [ 'calculatedfinishedsquarefeet',
 'taxvaluedollarcnt',
 'yearbuilt',
 'taxamount']
 
 - these columns are not helpful with value counts
 
 'yearbuilt' is kind of in the middle
 
 - do an nunique to see how many unique values are in each one


In [24]:
cols2 =  [ 'calculatedfinishedsquarefeet',
 'taxvaluedollarcnt',
 'yearbuilt',
 'taxamount']

In [29]:
# print unique values in columns
for col in cols:
    print(f'---------------')
    print(f'There are {df[col].nunique()} unique values in the {col} column')

---------------
There are 19 unique values in the bedroomcnt column
---------------
There are 38 unique values in the bathroomcnt column
---------------
There are 10580 unique values in the calculatedfinishedsquarefeet column
---------------
There are 592269 unique values in the taxvaluedollarcnt column
---------------
There are 153 unique values in the yearbuilt column
---------------
There are 918838 unique values in the taxamount column
---------------
There are 3 unique values in the fips column


### Let's deal with bedroomcnt and bathroom cnt first (11 in each)
- check and see if the bedroom count and bathroom count nulls are part of the same rows
- are those 11 NaN values in each one part of the same row

In [37]:
bedroom_count_nans = df[df.bedroomcnt.isnull() == True]
bedroom_count_nans

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
107763          NaN          NaN                           NaN   
118612          NaN          NaN                           NaN   
193993          NaN          NaN                        1348.0   
1141339         NaN          NaN                         200.0   
1324608         NaN          NaN                         990.0   
1442975         NaN          NaN                           NaN   
1647346         NaN          NaN                         400.0   
1701026         NaN          NaN                           NaN   
1722707         NaN          NaN                           NaN   
1776422         NaN          NaN                           NaN   
1797985         NaN          NaN                           NaN   

         taxvaluedollarcnt  yearbuilt  taxamount    fips  
107763             67366.0     1926.0     780.54  6059.0  
118612             43992.0     1946.0     541.64  6059.0  
193993            840698.0     1952.0        NaN  6059.0  
1141339           188972.0        NaN        NaN  6037.0  
1324608           435000.0     1906.0        NaN  6037.0  
1442975           273196.0        NaN        NaN  6037.0  
1647346            28347.0     1954.0        NaN  6037.0  
1701026           407930.0     1926.0        NaN  6037.0  
1722707           477161.0        NaN        NaN  6037.0  
1776422            38855.0        NaN        NaN  6037.0  
1797985          1017957.0     1990.0        NaN  6037.0

In [39]:
bedroom_count_nans.shape

(11, 7)

In [41]:
df[(df.bathroomcnt.isnull() == True) & (df.bedroomcnt.isnull() == True)]

bedroomcnt  bathroomcnt  calculatedfinishedsquarefeet  \
107763          NaN          NaN                           NaN   
118612          NaN          NaN                           NaN   
193993          NaN          NaN                        1348.0   
1141339         NaN          NaN                         200.0   
1324608         NaN          NaN                         990.0   
1442975         NaN          NaN                           NaN   
1647346         NaN          NaN                         400.0   
1701026         NaN          NaN                           NaN   
1722707         NaN          NaN                           NaN   
1776422         NaN          NaN                           NaN   
1797985         NaN          NaN                           NaN   

         taxvaluedollarcnt  yearbuilt  taxamount    fips  
107763             67366.0     1926.0     780.54  6059.0  
118612             43992.0     1946.0     541.64  6059.0  
193993            840698.0     1952.0        NaN  6059.0  
1141339           188972.0        NaN        NaN  6037.0  
1324608           435000.0     1906.0        NaN  6037.0  
1442975           273196.0        NaN        NaN  6037.0  
1647346            28347.0     1954.0        NaN  6037.0  
1701026           407930.0     1926.0        NaN  6037.0  
1722707           477161.0        NaN        NaN  6037.0  
1776422            38855.0        NaN        NaN  6037.0  
1797985          1017957.0     1990.0        NaN  6037.0

In [40]:
df[(df.bathroomcnt.isnull() == True) & (df.bedroomcnt.isnull() == True)].shape

(11, 7)

### Takeaways
- Bedroom and bathroom count both have Nans in the same columns
- decision: Let's drop those 11 columns
- not very much data in the grand scheme of things
- 2152863 rows to start with
- there should be 2152852 after subtracting 11 rows

In [43]:
# include this code in your wrangle function

df = df.drop(df[(df.bathroomcnt.isnull() == True) & (df.bedroomcnt.isnull() == True)].index)

# this one is a check don't include this in function
df.shape

(2152852, 7)

In [44]:
missing_values_table(df)

Your selected dataframe has 7 columns.
There are 4 columns that have missing values.


Missing Values  % of Total Values
yearbuilt                               9333                0.4
calculatedfinishedsquarefeet            8477                0.4
taxamount                               4433                0.2
taxvaluedollarcnt                        493                0.0